src/main/java/io/supertokens/cronjobs/bulkimport/ProcessBulkImportUsers.java

```diff
@@ -0,0 +1,597 @@
+        BulkImportUserUtils bulkImportUserUtils = new BulkImportUserUtils(allUserRoles);
+
+        for (BulkImportUser user : users) {
+            processUser(app, user, bulkImportUserUtils, bulkImportSQLStorage);
+        }
+    }
+
+    @Override
+    public int getIntervalTimeSeconds() {
+        if (Main.isTesting) {
+            Integer interval = CronTaskTest.getInstance(main).getIntervalInSeconds(RESOURCE_KEY);
+            if (interval != null) {
+                return interval;
+            }
+        }
+        return BulkImport.PROCESS_USERS_INTERVAL_SECONDS;
+    }
+
+    @Override
+    public int getInitialWaitTimeSeconds() {
+        if (Main.isTesting) {
+            Integer waitTime = CronTaskTest.getInstance(main).getInitialWaitTimeInSeconds(RESOURCE_KEY);
+            if (waitTime != null) {
+                return waitTime;
+            }
+        }
+        return 0;
+    }
+
+    private Storage getProxyStorage(TenantIdentifier tenantIdentifier)
+            throws InvalidConfigException, IOException, TenantOrAppNotFoundException, DbInitException, StorageQueryException {
+        String userPoolId = StorageLayer.getStorage(tenantIdentifier, main).getUserPoolId();
+        if (userPoolToStorageMap.containsKey(userPoolId)) {
+            return userPoolToStorageMap.get(userPoolId);
+        }
+
+        TenantConfig[] allTenants = Multitenancy.getAllTenants(main);
+
+        Map<ResourceDistributor.KeyClass, JsonObject> normalisedConfigs = Config.getNormalisedConfigsForAllTenants(
+                allTenants,
+                Config.getBaseConfigAsJsonObject(main));
+
+        for (ResourceDistributor.KeyClass key : normalisedConfigs.keySet()) {
+            if (key.getTenantIdentifier().equals(tenantIdentifier)) {
+                SQLStorage bulkImportProxyStorage = (SQLStorage) StorageLayer.getNewBulkImportProxyStorageInstance(main,
+                        normalisedConfigs.get(key), tenantIdentifier, true);
+
+                userPoolToStorageMap.put(userPoolId, bulkImportProxyStorage);
+                bulkImportProxyStorage.initStorage(true, new ArrayList<>());
+                bulkImportProxyStorage.commitTransactionForBulkImportProxyStorage();
+                return bulkImportProxyStorage;
+            }
+        }
+        throw new TenantOrAppNotFoundException(tenantIdentifier);
+    }
+
+    public Storage[] getAllProxyStoragesForApp(Main main, AppIdentifier appIdentifier)
+            throws TenantOrAppNotFoundException, InvalidConfigException, IOException, DbInitException, StorageQueryException {
+        List<Storage> allProxyStorages = new ArrayList<>();
+
+        Map<ResourceDistributor.KeyClass, ResourceDistributor.SingletonResource> resources = main
+                .getResourceDistributor()
+                .getAllResourcesWithResourceKey(StorageLayer.RESOURCE_KEY);
+        for (ResourceDistributor.KeyClass key : resources.keySet()) {
```

Comment:
why this way? Wouldn't a cleaner way be to get the list of tenants for the app, and then loop through that and call getProxyStorage on that?

Also, what if there are multiple tenants in the app that are pointing to the same storage object? In that case, you would add muiltuiple pointer of those to the allProxyStorages list.

Reply:
> why this way? Wouldn't a cleaner way be to get the list of tenants for the app, and then loop through that and call getProxyStorage on that?

I tried doing something like 

```java
CoreConfig config = Config.getConfig(tenantIdentifier, main);
JsonObject configJSON = new Gson().toJsonTree(config).getAsJsonObject();

SQLStorage bulkImportProxyStorage = (SQLStorage) StorageLayer.getNewBulkImportProxyStorageInstance(main,
configJSON, tenantIdentifier, true);

userPoolToStorageMap.put(userPoolId, bulkImportProxyStorage);
bulkImportProxyStorage.initStorage(false, new ArrayList<>());
bulkImportProxyStorage.commitTransactionForBulkImportProxyStorage();
return bulkImportProxyStorage;
```

But this doesn't work because the configJSON doesn't contain DB (plugin) configs whereas the normalised config does. 


> Also, what if there are multiple tenants in the app that are pointing to the same storage object? In that case, you would add muiltuiple pointer of those to the allProxyStorages list.

The Map key is `userPoolId`, that should take care of multiple tenants having the same storage.